databricks/sdk/mixins/workspace.py

```diff
@@ -1,14 +1,128 @@
+        self._notebooks_modified_after = notebooks_modified_after
+        self._scans = 0
+        self._in_progress = 0
+        # work queue is seeded from main thread and written/consumed from worker threads
+        self._work = Queue(self._threads)
+        # results queue is written from worker threads and consumed from the main thread
+        self._results = Queue(1)
+        # running condition guards self._in_progress, read from main/worker/reporter threads
+        # and modified by worker threads
+        self._cond = Condition()
+        self._reporter_interval = 60
+        self._reporter_cond = Condition()
+
+    def _enter_folder(self, path: str):
+        with self._cond:
+            _LOG.debug(f"Entering folder: {path}")
+            self._in_progress += 1
+        self._work.put(path)
+
+    def _leave_folder(self, path: str):
+        self._work.task_done()
+        with self._cond:
+            self._in_progress -= 1
+            self._scans += 1
+            _LOG.debug(f"Leaving folder: {path}")
+            if self._in_progress != 0:
+                return
+            self._cond.notify_all()
+            _LOG.debug("Sending poison pills to other workers")
+            for _ in range(0, self._threads - 1):
+                self._work.put(None)
+            with self._reporter_cond:
+                self._reporter_cond.notify()
+
+    def _is_running(self):
+        with self._cond:
+            return self._in_progress > 0
+
+    def _reporter(self):
+        _LOG.debug("Starting workspace listing reporter")
+        while self._is_running():
+            with self._reporter_cond:
+                self._reporter_cond.wait(self._reporter_interval)
+            scans = self._scans
+            took = datetime.now() - self._started
+            rps = int(scans / took.total_seconds())
+            _LOG.info(f"Scanned {scans} workspace folders at {rps}rps")
+        _LOG.debug("Finished workspace listing reporter")
+
+    def _worker(self, num):
+        _LOG.debug(f"Starting workspace listing worker {num}")
+        while self._is_running():
+            path = self._work.get()
+            if path is None:
+                self._work.task_done()
+                break # poison pill
+            try:
+                self._list(path)
+            except DatabricksError as err:
+                # See https://github.com/databrickslabs/ucx/issues/230
+                if err.error_code != "RESOURCE_DOES_NOT_EXIST":
+                    raise err
+                _LOG.warning(f"{path} is not listable. Ignoring")
+            finally:
+                self._leave_folder(path)
+        _LOG.debug(f"Finished workspace listing worker {num}")
+
+    def _list(self, path):
+        listing = self._listing(path, notebooks_modified_after=self._notebooks_modified_after)
+        for object_info in sorted(listing, key=lambda _: _.path):
+            if object_info.object_type != ObjectType.DIRECTORY:
```

Comment:
I think this should be `==`? The code in this block refers to what happens in a directory.